Tools/LocomotionTool/LocomotionTool.cs

```diff
@@ -0,0 +1,559 @@
+#if UNITY_EDITOR
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEditor.Experimental.EditorVR.Core;
+using UnityEditor.Experimental.EditorVR.Utilities;
+using UnityEngine;
+using UnityEngine.InputNew;
+using UnityEngine.UI;
+
+namespace UnityEditor.Experimental.EditorVR.Tools
+{
+	sealed class LocomotionTool : MonoBehaviour, ITool, ILocomotor, IUsesRayOrigin, ISetDefaultRayVisibility, ICustomActionMap,
+		ILinkedObject, IUsesViewerScale, ISettingsMenuItemProvider, ISerializePreferences
+	{
+		const float k_FastMoveSpeed = 20f;
+		const float k_SlowMoveSpeed = 1f;
+		const float k_RotationDamping = 0.2f;
+		const float k_RotationThreshold = 0.75f;
+
+		//TODO: Fix triangle intersection test at tiny scales, so this can go back to 0.01
+		const float k_MinScale = 0.1f;
+		const float k_MaxScale = 1000f;
+
+		const string k_WorldScaleProperty = "_WorldScale";
+
+		const int k_RotationSegments = 32;
+
+		enum State
+		{
+			Inactive = 0,
+			Aiming = 1,
+			Moving = 3
+		}
+
+		[SerializeField]
+		GameObject m_BlinkVisualsPrefab;
+
+		[SerializeField]
+		GameObject m_ViewerScaleVisualsPrefab;
+
+		[SerializeField]
+		ActionMap m_BlinkActionMap;
+
+		[SerializeField]
+		GameObject m_SettingsMenuItemPrefab;
+
+		[Serializable]
+		class Preferences
+		{
+			[SerializeField]
+			bool m_BlinkMode;
+
+			public bool blinkMode { get { return m_BlinkMode; } set { m_BlinkMode = value; } }
+		}
+
+		Preferences m_Preferences;
+
+		ViewerScaleVisuals m_ViewerScaleVisuals;
+
+		GameObject m_BlinkVisualsGO;
+		BlinkVisuals m_BlinkVisuals;
+
+		State m_State = State.Inactive;
+
+		bool m_AllowScaling = true;
+		bool m_Scaling;
+		float m_StartScale;
+		float m_StartDistance;
+		Vector3 m_StartPosition;
+		Vector3 m_StartMidPoint;
+		Vector3 m_StartDirection;
+		float m_StartYaw;
+
+		bool m_Rotating;
+		bool m_WasRotating;
+		bool m_Crawling;
+		Vector3 m_RayOriginStartPosition;
+		Vector3 m_RayOriginStartForward;
+		Vector3 m_RayOriginStartRight;
+		Quaternion m_RigStartRotation;
+		Vector3 m_RigStartPosition;
+		Vector3 m_CameraStartPosition;
+		Quaternion m_LastRotationDiff;
+
+		// Allow shared updater to check input values and consume controls
+		LocomotionInput m_LocomotionInput;
+
+		Camera m_MainCamera;
+		float m_OriginalNearClipPlane;
+		float m_OriginalFarClipPlane;
+
+		Toggle m_FlyToggle;
+		Toggle m_BlinkToggle;
+		bool m_BlockValueChangedListener;
+
+		public ActionMap actionMap { get { return m_BlinkActionMap; } }
+
+		public Transform rayOrigin { get; set; }
+
+		public Transform cameraRig { private get; set; }
+
+		public List<ILinkedObject> linkedObjects { private get; set; }
+
+		public GameObject settingsMenuItemPrefab { get { return m_SettingsMenuItemPrefab; } }
+
+		public GameObject settingsMenuItemInstance
+		{
+			set
+			{
+				foreach (var toggle in value.GetComponentsInChildren<Toggle>())
+				{
+					if (toggle.isOn)
```

Comment:
Yeah. It's basically a way to avoid having to create a new script to put on the prefab just for the purpose of labeling the toggles. I'm happy to do it that way, if you prefer, or come up with a better heuristic (say, first child or something)